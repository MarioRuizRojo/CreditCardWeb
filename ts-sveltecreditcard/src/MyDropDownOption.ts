/**
 * Option of the dropdown component
 */
export class MyDropDownOption {
    /**
     * Function to execute if option clicked
     */
    private handleClick: () => void;
    /**
     * Text to show in option's button
     */
    private text: string;
    /**
     * Css class of flag icon
     */
    private cssClass: string;

    /**
     * It throws error if Function to execute is undefined
     * It throws error if Text to show is undefined
     * It throws error if Css class is undefined
     * @param handCl Function to execute when click
     * @param txt Text to show in button
     * @param csscl Css class of flag icon
     */
    private checkAllNeededPropsCreateMyDropDownOption(handCl: () => void, txt: string, csscl: string): void {
        if (handCl == undefined || handCl == null) throw new Error("without handlerClickFunction cannot create MyDropDownOption");
        if (txt == null || txt == undefined) throw new Error("without captionTextDisplay cannot create MyDropDownOption");
        if (csscl == null || csscl == undefined) throw new Error("without cssClassString cannot create MyDropDownOption");
    }

    /**
     * It checks all parameters are not undefined and creates the object
     * @param handCl Function to execute when click
     * @param txt Text to show in button
     * @param csscl Css class of flag icon
     */
    constructor(handCl: () => void, txt: string, csscl: string) {
        this.checkAllNeededPropsCreateMyDropDownOption(handCl, txt, csscl);
        this.handleClick = handCl;
        this.text = txt;
        this.cssClass = csscl;
    }

    /**
     * 
     * @returns Function to execute if option clicked
     */
    public getHandleClick(): () => void {
        return this.handleClick;
    }

    /**
     * 
     * @param handleclick Function to execute to set
     */
    public setHandleClick(handleclick: () => void): void {
        this.handleClick = handleclick;
    }

    /**
     * 
     * @returns Text to show in option's button
     */
    public getText(): string {
        return this.text;
    }

    /**
     * 
     * @param text Text displayed to set
     */
    public setText(text: string): void {
        this.text = text;
    }

    /**
     * 
     * @returns Css class of flag icon
     */
    public getCssClass(): string {
        return this.cssClass;
    }

    /**
     * 
     * @param cssClass Css class of flag icon to set
     */
    public setCssClass(cssClass: string): void {
        this.cssClass = cssClass;
    }
}