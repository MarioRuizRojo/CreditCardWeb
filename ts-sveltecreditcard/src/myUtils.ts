import { iconClassBase, iconClassGreen, iconClassRed } from "./constants";

/**
 * It returns red color clases if the validation fails and green ones if it succeeds
 * @param passed validation status
 * @param iconClassX css base class of icon
 * @returns css classes of icon
 */
export function iconSuccedOrFailed(passed: boolean, iconClassX: string): string {
    if (passed) {
        return iconClassBase + iconClassX + iconClassGreen;
    }
    else {
        return iconClassBase + iconClassX + iconClassRed;
    }
}

/**
 * Very useful dummy function that has no parameters and does nothing
 */
export function dummyFunction(): void {
    //do nothing
}