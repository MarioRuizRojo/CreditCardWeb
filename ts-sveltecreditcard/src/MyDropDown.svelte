<script lang="ts">
  import { onMount } from "svelte";

  import { iconClassCaretDown, iconClassCaretLeft } from "./constants";
  import type { MyDropDownOption } from "./MyDropDownOption";
  import { dummyFunction } from "./myUtils";
  /**
   * List of options to show inside the dropdown
   */
  export let dropDownOptions: MyDropDownOption[] = [];
  /**
   * label caption text to show in the main button of dropdown
   */
  export let mainCaption: String = "";

  /**
   * list of options of dropdown hidden if true
   */
  let hideOptionsList: boolean = true;
  /**
   * caret icon to show, left caret if the options list is hidden
   * and down caret if the options list is shown
   */
  let caret: string = iconClassCaretLeft;
  /**
   * Height of the options list div container
   */
  let optionsDivHeight: number = 0;
  /**
   * X axis position of the options list div container
   * it is a relative position to main button
   */
  let optionsDivPositionTop: number = 0;

  /**
   * When main button mounted it calculates proportional size and
   * position of the options list div container
   */
  onMount(function (): void {
    let mainButton: HTMLElement = document.getElementById("mainButton");
    let mainButtonHeight: number = mainButton.offsetHeight;
    optionsDivHeight = 1.6 * mainButtonHeight;
    optionsDivPositionTop = -0.3 * mainButtonHeight;
  });

  /**
   * It alternate visibility of the options list div container
   * and the caret icon direction
   * @param handleOption function to execute if the option is clicked
   */
  function handleClick(handleOption: () => void): void {
    hideOptionsList = !hideOptionsList;
    if (caret == iconClassCaretLeft) {
      caret = iconClassCaretDown;
    } else {
      caret = iconClassCaretLeft;
    }
    handleOption(); //if clicked from main button this function is dummy ()=>{}
  }
</script>

<div class="col">
  <!-- Main Button -->
  <button
    id="mainButton"
    data-testid="mainbuttondropdown"
    class="row dropdownred"
    type="button"
    on:click={() => handleClick(dummyFunction)}
  >
    {mainCaption}
    <!-- Caret -->
    <svg
      xmlns="http://www.w3.org/2000/svg"
      width="16"
      height="16"
      fill="currentColor"
      class={"iconCaret " + caret}
      viewBox="0 0 16 16"
    >
      {#if hideOptionsList}
        <path
          d="M10 12.796V3.204L4.519 8 10 12.796zm-.659.753-5.48-4.796a1 1 0 0 1 0-1.506l5.48-4.796A1 1 0 0 1 11 3.204v9.592a1 1 0 0 1-1.659.753z"
        />
      {:else}
        <path
          d="M3.204 5h9.592L8 10.481 3.204 5zm-.753.659 4.796 5.48a1 1 0 0 0 1.506 0l4.796-5.48c.566-.647.106-1.659-.753-1.659H3.204a1 1 0 0 0-.753 1.659z"
        />
      {/if}
    </svg>
  </button>
  <!-- options div -->
  <div
    class="row"
    style={"display: inline-block; position: absolute; inset: " +
      optionsDivPositionTop +
      "px auto auto 0px; margin: 0px; transform: translate(0px, " +
      optionsDivHeight +
      "px);"}
    hidden={hideOptionsList}
  >
    {#each dropDownOptions as dropDownOption, index}
      <!-- option -->
      <button
        data-testid={"option" + index}
        class="row btn dropdownred"
        type="button"
        on:click={() => handleClick(dropDownOption.getHandleClick())}
      >
        <i class={dropDownOption.getCssClass()} />{dropDownOption.getText()}
      </button>
    {/each}
  </div>
</div>

<style>
  .dropdownred {
    color: grey;
    background-color: #540402;
  }
  .iconCaret {
    display: inline-block;
    min-width: 16px;
    height: 16px;
    padding-right: 0px;
    padding-left: 0px;
    max-width: 16px;
    vertical-align: middle;
    line-height: revert;
    position: relative;
    top: 5px;
  }
</style>
