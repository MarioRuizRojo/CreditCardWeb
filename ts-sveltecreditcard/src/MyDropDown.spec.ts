import MyDropDown from './MyDropDown.svelte';
import { render, fireEvent } from '@testing-library/svelte';

import { MyDropDownOption } from './MyDropDownOption';
import { dummyFunction } from './myUtils';
/**
 * Mocked function to execute when option clicked
 */
let mockHandl = jest.fn();
/**
 * Option 1 as test data set
 */
let dropDownOption1: MyDropDownOption = new MyDropDownOption(dummyFunction, "Option 1", "flag flag-uk");
/**
 * Option 2 as test data set
 */
let dropDownOption2: MyDropDownOption = new MyDropDownOption(mockHandl, "Option 2", "flag flag-es");
/**
 * options list as test data set
 */
let dropDownOptions: MyDropDownOption[] = [dropDownOption1, dropDownOption2];
let props: any = {
    dropDownOptions: dropDownOptions,
    mainCaption: "main text caption display"
};

/**
 * It expects test data set values for MyDropDown attributes
 * and snapshot of no selected option in MyDropDown
 */
it('check constructor', function (): void {
    const rendered = render(MyDropDown, props);
    expect(rendered).toMatchSnapshot();
});

/**
 * It expects mocked handler click function to be called when user click the option
 * and snapshot of that option selected in MyDropDown
 */
it('check handler click function', async function (): Promise<void> {
    const rendered = render(MyDropDown, props);
    const mainButton = await rendered.findByTestId("mainbuttondropdown");
    await fireEvent.click(mainButton);
    const option1Button = await rendered.findByTestId("option1");
    await fireEvent.click(option1Button);
    expect(mockHandl).toHaveBeenCalled();
    expect(rendered).toMatchSnapshot();
});