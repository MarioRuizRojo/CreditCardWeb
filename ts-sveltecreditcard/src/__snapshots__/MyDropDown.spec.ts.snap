// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`check constructor 1`] = `
Object {
  "component": MyDropDown {
    "$$": Object {
      "after_update": Array [],
      "before_update": Array [],
      "bound": Object {},
      "callbacks": Object {},
      "context": Map {},
      "ctx": Array [
        Array [
          MyDropDownOption {
            "cssClass": "flag flag-uk",
            "handleClick": [Function],
            "text": "Option 1",
          },
          MyDropDownOption {
            "cssClass": "flag flag-es",
            "handleClick": [MockFunction],
            "text": "Option 2",
          },
        ],
        "main text caption display",
        true,
        "bi bi-caret-left",
        0,
        -0,
        [Function],
        [Function],
        [Function],
      ],
      "dirty": Array [
        -1,
      ],
      "fragment": Object {
        "c": [Function],
        "d": [Function],
        "i": [Function],
        "l": [Function],
        "m": [Function],
        "o": [Function],
        "p": [Function],
      },
      "not_equal": [Function],
      "on_destroy": Array [
        [Function],
      ],
      "on_disconnect": Array [],
      "on_mount": Array [],
      "props": Object {
        "dropDownOptions": 0,
        "mainCaption": 1,
      },
      "skip_bound": false,
      "update": [Function],
    },
    "$$set": [Function],
    "$capture_state": [Function],
    "$inject_state": [Function],
  },
  "container": <body>
    <div>
      <div
        class="col"
      >
        <button
          class="row dropdownred svelte-2puf0x"
          data-testid="mainbuttondropdown"
          id="mainButton"
          type="button"
        >
          main text caption display
           
          <svg
            class="iconCaret bi bi-caret-left svelte-2puf0x"
            fill="currentColor"
            height="16"
            viewBox="0 0 16 16"
            width="16"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M10 12.796V3.204L4.519 8 10 12.796zm-.659.753-5.48-4.796a1 1 0 0 1 0-1.506l5.48-4.796A1 1 0 0 1 11 3.204v9.592a1 1 0 0 1-1.659.753z"
            />
          </svg>
        </button>
         
        <div
          class="row"
          hidden=""
          style="display: inline-block; position: absolute; inset: 0px auto auto 0px; margin: 0px; transform: translate(0px, 0px);"
        >
          <button
            class="row btn dropdownred svelte-2puf0x"
            data-testid="option0"
            type="button"
          >
            <i
              class="flag flag-uk svelte-2puf0x"
            />
            Option 1
             
          </button>
          <button
            class="row btn dropdownred svelte-2puf0x"
            data-testid="option1"
            type="button"
          >
            <i
              class="flag flag-es svelte-2puf0x"
            />
            Option 2
             
          </button>
        </div>
      </div>
    </div>
  </body>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;

exports[`check handler click function 1`] = `
Object {
  "component": MyDropDown {
    "$$": Object {
      "after_update": Array [],
      "before_update": Array [],
      "bound": Object {},
      "callbacks": Object {},
      "context": Map {},
      "ctx": Array [
        Array [
          MyDropDownOption {
            "cssClass": "flag flag-uk",
            "handleClick": [Function],
            "text": "Option 1",
          },
          MyDropDownOption {
            "cssClass": "flag flag-es",
            "handleClick": [MockFunction] {
              "calls": Array [
                Array [],
              ],
              "results": Array [
                Object {
                  "type": "return",
                  "value": undefined,
                },
              ],
            },
            "text": "Option 2",
          },
        ],
        "main text caption display",
        true,
        "bi bi-caret-left",
        0,
        -0,
        [Function],
        [Function],
        [Function],
      ],
      "dirty": Array [
        -1,
      ],
      "fragment": Object {
        "c": [Function],
        "d": [Function],
        "i": [Function],
        "l": [Function],
        "m": [Function],
        "o": [Function],
        "p": [Function],
      },
      "not_equal": [Function],
      "on_destroy": Array [
        [Function],
      ],
      "on_disconnect": Array [],
      "on_mount": Array [],
      "props": Object {
        "dropDownOptions": 0,
        "mainCaption": 1,
      },
      "skip_bound": false,
      "update": [Function],
    },
    "$$set": [Function],
    "$capture_state": [Function],
    "$inject_state": [Function],
  },
  "container": <body>
    <div>
      <div
        class="col"
      >
        <button
          class="row dropdownred svelte-2puf0x"
          data-testid="mainbuttondropdown"
          id="mainButton"
          type="button"
        >
          main text caption display
           
          <svg
            class="iconCaret bi bi-caret-left svelte-2puf0x"
            fill="currentColor"
            height="16"
            viewBox="0 0 16 16"
            width="16"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M10 12.796V3.204L4.519 8 10 12.796zm-.659.753-5.48-4.796a1 1 0 0 1 0-1.506l5.48-4.796A1 1 0 0 1 11 3.204v9.592a1 1 0 0 1-1.659.753z"
            />
          </svg>
        </button>
         
        <div
          class="row"
          hidden=""
          style="display: inline-block; position: absolute; inset: 0px auto auto 0px; margin: 0px; transform: translate(0px, 0px);"
        >
          <button
            class="row btn dropdownred svelte-2puf0x"
            data-testid="option0"
            type="button"
          >
            <i
              class="flag flag-uk svelte-2puf0x"
            />
            Option 1
             
          </button>
          <button
            class="row btn dropdownred svelte-2puf0x"
            data-testid="option1"
            type="button"
          >
            <i
              class="flag flag-es svelte-2puf0x"
            />
            Option 2
             
          </button>
        </div>
      </div>
    </div>
  </body>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
